pkg/sync/file/filepath_sync.go

```diff
@@ -22,31 +22,49 @@ type Sync struct {
 	ProviderArgs sync.ProviderArgs
 	// FileType indicates the file type e.g., json, yaml/yml etc.,
 	fileType string
+	watcher  *fsnotify.Watcher
+	ready    bool
+	Mux      *msync.RWMutex
 }
 
 // default state is used to prevent EOF errors when handling filepath delete events + empty files
 const defaultState = "{}"
 
-//nolint:funlen
-func (fs *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
+func (fs *Sync) Init(ctx context.Context) error {
 	fs.Logger.Info("Starting filepath sync notifier")
-	watcher, err := fsnotify.NewWatcher()
+	w, err := fsnotify.NewWatcher()
 	if err != nil {
 		return err
 	}
-	defer watcher.Close()
-
-	err = watcher.Add(fs.URI)
+	fs.watcher = w
+	err = fs.watcher.Add(fs.URI)
 	if err != nil {
 		return err
 	}
+	return nil
+}
 
-	fs.sendDataSync(ctx, sync.ALL, dataSync)
+func (fs *Sync) IsReady() bool {
+	fs.Mux.RLock()
```

Comment:
Is the mutex strictly necessary if only one goroutine can change the state?

Reply:
I added this because there were complains on race-conditions while running the tests